src/os/bluestore/BlueStore.cc

```diff
@@ -9072,34 +9378,84 @@ int BlueStore::_fsck_on_open(BlueStore::FSCKDepth depth, bool repair)
         }
       }
       fm->enumerate_reset();
-      size_t count = used_blocks.count();
-      if (used_blocks.size() != count) {
-        ceph_assert(used_blocks.size() > count);
-        used_blocks.flip();
-        size_t start = used_blocks.find_first();
-        while (start != decltype(used_blocks)::npos) {
-	  size_t cur = start;
-	  while (true) {
-	    size_t next = used_blocks.find_next(cur);
-	    if (next != cur + 1) {
-	      ++errors;
-	      derr << "fsck error: leaked extent 0x" << std::hex
-		   << ((uint64_t)start * fm->get_alloc_size()) << "~"
-		   << ((cur + 1 - start) * fm->get_alloc_size()) << std::dec
-		   << dendl;
-	      if (repair) {
-	        repairer.fix_leaked(db,
-				    fm,
-				    start * min_alloc_size,
-				    (cur + 1 - start) * min_alloc_size);
+
+#ifdef HAVE_LIBZBD
+      if (freelist_type == "zoned") {
+	// verify per-zone state: count "dead" blocks in zone (neither allocator nor
+	// free space past the write pointer)
+	auto a = dynamic_cast<ZonedAllocator*>(alloc);
+	used_blocks.flip();
+	auto num_zones = bdev->get_size() / zone_size;
+
+	uint64_t pos = used_blocks.find_first();
+
+	// skip conventional zones
+	while (pos != decltype(used_blocks)::npos &&
+	       (pos * min_alloc_size) / zone_size < first_sequential_zone) {
+	  dout(40) << " (conventional) zone 0x" << std::hex
+		   << (pos * min_alloc_size) / zone_size
+		   << " dead 0x" << (pos * min_alloc_size) << "~" << min_alloc_size
+		   << std::dec << dendl;
+	  pos = used_blocks.find_next(pos);
```

Comment:
I see ZonedAllocator is allocating only from sequential zones.
Should presence of used_blocks in conventional zones be considered an error?

Reply:
Perhaps, but I don't want to invest too much effort here since I think this is likely to change.  (1) There are partial bluefs changes that allow it to consume sequential zones. (2) We probably want bluefs to spill over into sequential zones as needed for sst files, but use the conventional region for its log files.  (3) We could conceivably have bluestore put some objects in the conventional zone if the hints indicate a high likelihood of overwrites.  If we go down any of these paths, then we probably need a single Allocator that will pull from one region or the other based on some flag.